### ISSUE TITLE
`uv pip compile` causes "direct URL requirements without package names are not supported" error

### ISSUE THREAD

**Issue by groodt** [astral-sh/uv#3818]
It appears that there is something strange happening when `uv pip compile` is creating a lockfile for direct url references when environment variable interpolation is occurring.

## requirements.in
```
requests @ ${GITHUB_URL}/psf/requests
```

## Create lockfile
```
GITHUB_URL=git+https://github.com python -m uv pip compile requirements.in --ou
tput-file requirements.txt
```

## Inspect generated requirements.txt
```
# This file was autogenerated by uv via the following command:
#    uv pip compile requirements.in --output-file requirements.txt
certifi==2024.2.2
    # via requests
charset-normalizer==3.3.2
    # via requests
idna==3.7
    # via requests
${GITHUB_URL}/psf/requests@88dce9d854797c05d0ff296b70e0430535ef8aaf
    # via -r requirements.in
urllib3==2.2.1
    # via requests
```

## Run lockfile generation again and an error occurs
```
GITHUB_URL=git+https://github.com python -m uv pip compile requirements.in --ou
tput-file requirements.txt
error: direct URL requirements without package names are not supported: `git+https://github.com/psf/requests@88dce9d854797c05d0ff296b70e0430535ef8aaf`
```

**Comment by charliermarsh** [comment]
I’ll try to get this out today, I think it’s a regression from a recent change.

**Comment by charliermarsh** [comment]
This should be fixed in v0.2.3 (out now).
